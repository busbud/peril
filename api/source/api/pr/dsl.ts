import * as express from "express"
import winston from "../../logger"

import { GitHub } from "danger/distribution/platforms/GitHub"
import inline from "danger/distribution/runner/runners/inline"

import { getTemporaryAccessTokenForInstallation } from "../../api/github"
import { RunType } from "../../danger/danger_run"
import { executorForInstallation } from "../../danger/danger_runner"
import { getPerilPlatformForDSL } from "../../danger/peril_platform"
import { githubAPIForCommentable } from "../../github/events/utils/commenting"
import { PERIL_ORG_INSTALLATION_ID } from "../../globals"

export const prDSLRunner = async (req: express.Request, res: express.Response, _: express.NextFunction) => {
  winston.info("[router] -- Received OK")

  const query = req.query
  if (!query.owner) {
    return res.status(422).jsonp({ error: "No `owner` query param sent." })
  }

  if (!query.repo) {
    return res.status(422).jsonp({ error: "No `repo` query param sent." })
  }

  if (!query.number) {
    return res.status(422).jsonp({ error: "No `number` query param sent." })
  }

  // This has to be set for public usage.
  if (!PERIL_ORG_INSTALLATION_ID) {
    throw new Error("You can't support PR DSLs without setting up the PERIL_ORG_INSTALLATION_ID")
  }

  const token = await getTemporaryAccessTokenForInstallation(PERIL_ORG_INSTALLATION_ID)

  const ghDetails = {
    fullName: query.owner + "/" + query.repo,
    prID: query.number,
  }

  const githubAPI = githubAPIForCommentable(token, ghDetails.fullName, ghDetails.prID)

  const gh = GitHub(githubAPI)
  const platform = getPerilPlatformForDSL(RunType.pr, gh, {})

  const exec = await executorForInstallation(platform, inline, {})
  const dangerDSL = await exec.dslForDanger()

  // Remove this to reduce data
  if (dangerDSL.github) {
    dangerDSL.github.api = {} as any
  }

  // TODO: include Danger version number in JSON
  return res.status(400).jsonp({
    danger: dangerDSL,
    status: "OK",
  })
}
