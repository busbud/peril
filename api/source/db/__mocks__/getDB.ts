import { DatabaseAdaptor, GitHubInstallation } from ".."
import { RuntimeEnvironment } from "../runtimeEnv"

// Caches per test file, so you can import it and check from the import
// If this is an issue, don't use this mock :D
let perTestFileMock: MockDB

// A set of type alias' to make the below code readable.
type NullFunc = () => Promise<void>
type InstallationIDToNullInstallation = (installationID: number) => Promise<GitHubInstallation | null>
type InstallationIDToInstallation = (installationID: number) => Promise<GitHubInstallation>
type InstallationIDsToInstallations = (installationID: number[]) => Promise<GitHubInstallation[]>
type NoArgsToInstallations = () => Promise<GitHubInstallation[]>
type InstallationToVoid = (installationID: number) => Promise<void>

// Take the DB and rewrite all of its functions to be both the original version and potentially the
// jest mocked version of it.
export interface MockDB extends DatabaseAdaptor {
  setup: NullFunc & jest.Mock<NullFunc>
  getInstallation: InstallationIDToNullInstallation & jest.Mock<InstallationIDToNullInstallation>
  getInstallations: InstallationIDsToInstallations & jest.Mock<InstallationIDsToInstallations>
  updateInstallation: InstallationIDToNullInstallation & jest.Mock<InstallationIDToNullInstallation>
  saveInstallation: InstallationIDToInstallation & jest.Mock<InstallationIDToNullInstallation>
  deleteInstallation: InstallationToVoid & jest.Mock<InstallationToVoid>
  getSchedulableInstallationsWithKey: NoArgsToInstallations & jest.Mock<NoArgsToInstallations>
  clear: () => void
}

// Returns a MockDB which you'll have to alias to alas, because the type system doesn't know
// that jest's mocking system will return a MockDB instead of a typical DatabaseAdaptor

export const getDB = (): MockDB => {
  if (perTestFileMock) {
    return perTestFileMock
  }

  // Official API
  perTestFileMock = {
    getInstallation: jest.fn(),
    getInstallations: jest.fn(),
    deleteInstallation: jest.fn(),
    saveInstallation: jest.fn(),
    updateInstallation: jest.fn(),
    getSchedulableInstallationsWithKey: jest.fn(),
    setup: jest.fn(),
    clear: () => "void",
  }
  // So we can access the internals
  perTestFileMock.clear = () => {
    perTestFileMock.getInstallation.mockClear()
    perTestFileMock.getInstallations.mockClear()
    perTestFileMock.deleteInstallation.mockClear()
    perTestFileMock.saveInstallation.mockClear()
    perTestFileMock.updateInstallation.mockClear()
    perTestFileMock.getSchedulableInstallationsWithKey.mockClear()
    perTestFileMock.setup.mockClear()
  }

  return perTestFileMock
}

export const runtimeEnv = RuntimeEnvironment.Standalone
