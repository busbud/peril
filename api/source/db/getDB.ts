import { jsonDatabase } from "./json"
import { mongoDatabase } from "./mongo"

import { DatabaseAdaptor } from "."
import { RuntimeEnvironment } from "./runtimeEnv"

const isJest = typeof jest !== "undefined"

const hasJSONDef = !!process.env.DATABASE_JSON_FILE
const hasPerilAPIURL = !!process.env.PUBLIC_API_ROOT_URL

const hasLambdaEnv = !!process.env._HANDLER

/** There are three runtime environments for Peril, this says which one it is */
export const runtimeEnvironment = hasJSONDef
  ? RuntimeEnvironment.Standalone
  : hasPerilAPIURL
    ? RuntimeEnvironment.Peril
    : hasLambdaEnv
      ? RuntimeEnvironment.Runner
      : RuntimeEnvironment.Unknown

const getDatabaseForEnv = (env: any): DatabaseAdaptor | null => {
  if (env.DATABASE_JSON_FILE || isJest) {
    const json = jsonDatabase(env.DATABASE_JSON_FILE)
    json.setup()
    return json
  }

  if (env.MONGODB_URI) {
    if (!isJest) {
      mongoDatabase.setup()
    }
    return mongoDatabase
  }

  return null
}

let db: DatabaseAdaptor | null = null
/** Gets the Current DB for this runtime environment */
export const getDB = () => {
  if (!db) {
    db = getDatabaseForEnv(process.env)
  }

  if (!db) {
    throw new Error("No default DB was set up")
  }

  return db
}
