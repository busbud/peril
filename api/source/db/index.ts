import { GitHubInstallationSettings } from "./GitHubRepoSettings"

/**
 * Should look like one of the following:
 * - "dangerfile.js"
 * - "/path/to/dangerfile.ts"
 * - "repo/slug@to/dangerfile.ts"
 */
export type DangerfileReferenceString = string

/**
 * An event string would be something like:
 * - "pull_request.*"
 * - "pull_request.updated"
 * - "ping"
 * - "issue.*"
 */
export type PerilEventString = string

/** An individual integration of Danger via Peril, this is like the org */
export interface GitHubInstallation extends PerilInstallationSettings {
  // Not these settings are things added by Peril, and it's not expected
  // that they are in the settings.json nor available to JSON-based hosts

  /** The associated name of the installation owner */
  login: string
  /** An image url for the installation owner */
  avatarURL: string
  /** An image url for the installation owner */
  envVars?: any
  /** AWS function ID */
  lambdaName: string
}

export interface PerilSettingsRepoJSON {
  /**
   * In our DB this is represented as a JSON type, so you should anticipate have settings
   * as a nullable type. These are the entire installation settings.
   */
  settings: GitHubInstallationSettings

  /** Having rules in here would mean that it would happen on _any_ event, another JSON type in the DB */
  rules: RunnerRuleset

  /**
   * Scheduled tasks to run using a human readable syntax. This runs from a set of standard hooks
   * which will trigger running a task. The current hooks (may be out of date, see `InstallationSchedulerKeys` in
   * the codebase) are "hourly", "daily", "weekly", "monday-morning-est", "tuesday-morning-est",
   * "wednesday-morning-est", "thursday-morning-est", "friday-morning-est"
   *
   *     "scheduler" : {
   *       "daily": "daily-license-check",
   *       "weekly": "cleanup-stale-issues"
   *     }
   */
  scheduler: TaskObject

  /**
   * Individual tasks which a Peril can schedule, either via the Dangerfile API or via the
   * scheduler object. These keys are used by the scheduler in the settings JSON, and can be used
   * to trigger a job to occur in the future via `peril.scheduleTask` in a Dangerfile.
   *
   *   "tasks" : {
   *     "message-slack-dev-channel": "tasks/slackDevChannel.ts",
   *     "daily-license-check": "tasks/dailyLicenseCheck.ts",
   *     "standup": ["tasks/checkForOpenRFCs.ts", "tasks/checkForNewRepos.ts"],
   *   }
   */
  tasks: RunnerRuleset

  /**
   * A set of repos and their additional event hooks, these are
   * in addition to the ones provided by `"rules"` which are applied
   * to every repo.
   *
   *     "repos" : {
   *       "orta/ORStackView": {
   *          "issue.created": "orta/peril@lock_issues.ts"
   *       }
   *     }
   *
   */
  repos: UniqueRepoRuleset
}

/** An individual integration of Danger via Peril, this is like the org */
export interface PerilInstallationSettings extends PerilSettingsRepoJSON {
  /**
   * The ID Integration, this is used when talking to GitHub mainly, but is used
   * as a unique ID in our db
   */
  iID: number

  /**
   * The path to the settings repo and json file
   * e.g. danger/peril-settings@settings.json
   *
   * Filled in automatically when using JSON db, and
   * is the initial string when working in the public mode
   */
  perilSettingsJSONURL: DangerfileReferenceString

  /**
   * Slack URL that Peril can send system messages to
   */
  installationSlackUpdateWebhookURL: string | undefined
}

export interface UniqueRepoRuleset {
  [name: string]: RunnerRuleset
}

/** The available keys for scheduling a task against */
export type InstallationSchedulerKeys =
  | "hourly"
  | "daily"
  | "weekly"
  | "monday-morning-est"
  | "tuesday-morning-est"
  | "wednesday-morning-est"
  | "thursday-morning-est"
  | "friday-morning-est"

type TaskObject = { [P in InstallationSchedulerKeys]?: string }

export interface RunnerRuleset {
  [name: string]: DangerfileReferenceString | DangerfileReferenceString[]
}

export interface GithubRepo {
  /** The installation that this repo is connected to */
  installationID: number
  /** The full name of the slug for this repo, note these could not be unique (GitHub Enterprise?) */
  fullName: string
  /** Runner rules ID reference, another JSON type in the DB */
  rules: RunnerRuleset
}

export interface DatabaseAdaptor {
  /** A once per server start setup function */
  setup: () => Promise<void>

  /** Gets an integrations settings */
  getInstallation: (installationID: number) => Promise<GitHubInstallation | null>
  /** Updates an integrations settings */
  updateInstallation: (installationID: number) => Promise<GitHubInstallation | null>
  /** Saves an Integration */
  saveInstallation: (installation: GitHubInstallation) => Promise<GitHubInstallation>
  /** Deletes the operation */
  deleteInstallation: (installationID: number) => Promise<void>
  /** Gets an integrations settings */
  getSchedulableInstallationsWithKey: (key: string) => Promise<GitHubInstallation[]>
}
