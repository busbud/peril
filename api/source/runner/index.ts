import * as getSTDIN from "get-stdin"
import logger from "../logger"
import { run } from "./run"

/// This is only used by "runFromSameHost.ts"

try {
  // Provide a timeout mechanism for the STDIN from the hyper func host
  let foundDSL = false
  getSTDIN().then(stdin => {
    foundDSL = true
    run(stdin)
  })

  process.on("unhandledRejection", (error: Error) => {
    logger.error("unhandledRejection:", error.message, error.stack)
    process.exitCode = 1
  })

  // Add a timeout so that CI doesn't run forever if something has broken.
  setTimeout(() => {
    if (!foundDSL) {
      logger.error("Timeout: Failed to get the Peril DSL after 5 seconds")
      process.exitCode = 1
      process.exit(1)
    }
  }, 5000)
} catch (error) {
  const err = error as Error
  logger.error(`Error ${err.name} in the runner: ${err.message}\n${err.stack}`)
  throw error
}
