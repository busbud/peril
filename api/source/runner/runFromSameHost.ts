import { spawn, spawnSync } from "child_process"
import debug from "debug"
import { InstallationToRun } from "../danger/danger_runner"
import { sendSlackMessageToInstallationID } from "../infrastructure/installationSlackMessaging"
import { PerilRunnerBootstrapJSON } from "./triggerSandboxRun"

const d = debug("runFromSameHost")

export const runFromSameHost = async (
  stdOUT: PerilRunnerBootstrapJSON,
  // tslint:disable-next-line:variable-name
  _eventName: string,
  installation: InstallationToRun
) => {
  const which = spawnSync("/usr/bin/which", ["node"])
  const nodePath = which.stdout.toString().trim()

  const path = "out/runner/index.js"
  const child = spawn(nodePath, [path], { env: { ...process.env, ...stdOUT.perilSettings.envVars } })

  // Pipe in the STDOUT
  child.stdin.write(JSON.stringify(stdOUT))
  child.stdin.end()

  let allLogs = ""
  child.stdout.on("data", async data => {
    const stdout = data.toString()
    allLogs += stdout
    d(stdout)
  })

  child.stderr.on("data", data => {
    const stderr = data.toString()
    allLogs += stderr
    d(stderr)
  })

  child.on("close", async code => {
    d(`child process exited with code ${code}`)
    if (code) {
      sendSlackMessageToInstallationID(allLogs, installation.iID)
    }
  })
}
