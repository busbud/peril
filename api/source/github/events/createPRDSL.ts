import { jsonDSLGenerator } from "danger/distribution/runner/dslGenerator"
import { jsonToDSL } from "danger/distribution/runner/jsonToDSL"
import { getPerilPlatformForDSL } from "../../danger/peril_platform"

import { GitHub } from "danger/distribution/platforms/GitHub"
import { GitHubAPI } from "danger/distribution/platforms/github/GitHubAPI"
import { RunType } from "../../danger/danger_run"
import { source } from "../../danger/peril_ci_source"

/**
 * Generates a full DSL for a PR
 *
 * @param githubAPI the Danger GithubAPI instance
 */
export const createPRDSL = async (githubAPI: GitHubAPI) => {
  const jsonDSL = await createPRJSONDSL(githubAPI)
  // Danger JS expects the Github access token to be set on
  // settings.github.accessToken, so put it there if we have one
  if (githubAPI.token !== undefined) {
    jsonDSL.settings.github.accessToken = githubAPI.token
  }
  return await jsonToDSL(jsonDSL, source)
}

/**
 * Generates a full DSL for a PR
 *
 * @param githubAPI the Danger GithubAPI instance
 */
export const createPRJSONDSL = async (githubAPI: GitHubAPI) => {
  const gh = GitHub(githubAPI)
  const platform = getPerilPlatformForDSL(RunType.pr, gh, {})
  // These are what Danger JS uses to pass info to sub-commands
  // peril scopes all of its settings elsewhere, so a blank is fine
  const cliArgs = {} as any
  return await jsonDSLGenerator(platform, source, cliArgs)
}
