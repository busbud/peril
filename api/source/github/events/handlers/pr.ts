import { DangerDSLType } from "danger/distribution/dsl/DangerDSL"
import { DangerResults } from "danger/distribution/dsl/DangerResults"
import { DangerRun, RunType } from "../../../danger/danger_run"
import { runDangerForInstallation } from "../../../danger/danger_runner"
import { canUserWriteToRepo, doesPRExist, getGitHubFileContents } from "../../lib/github_helpers"
import { createPRDSL } from "../createPRDSL"
import { GitHubRunSettings } from "../github_runner"
import { Pull_request } from "../types/pull_request_updated.types"
import { githubAPIForCommentable } from "../utils/commenting"

export const runPRRun = async (
  eventName: string,
  runs: DangerRun[],
  settings: GitHubRunSettings,
  token: string,
  pr: Pull_request
): Promise<DangerResults | null> => {
  if (!runs.length) {
    return null
  }

  if (!settings.repoName) {
    console.error("An event without a repo name was passed to runRPRun") // tslint:disable-line
    return null
  }

  if (!settings.triggeredByUsername) {
    console.error("An event without a username was passed to runRPRun") // tslint:disable-line
    return null
  }

  if (!pr.head.repo) {
    console.error("An event without a head repo was passed to runPRRun") // tslint:disable-line
    return null
  }

  // No joke, GitHub's been sending up PRs which don't exist
  const exists = await doesPRExist(token, pr.base.repo.full_name, pr.number)
  if (!exists) {
    console.error("Can't run Danger against a PR which doesn't exist.") // tslint:disable-line
    return null
  }

  const githubAPI = githubAPIForCommentable(token, settings.repoName, settings.commentableID)

  const errorResults = await validateRuns(runs, settings, token, pr)
  if (errorResults) {
    return errorResults
  }

  const dangerfileReferences: string[] = []
  const contents: string[] = []
  for (const run of runs) {
    const dangerfileRepoForPR = pr.head.repo.full_name
    const repoForDangerfile = run.repoSlug || dangerfileRepoForPR

    const headDangerfile = await getGitHubFileContents(token, repoForDangerfile, run.dangerfilePath, run.branch)
    contents.push(headDangerfile)
    dangerfileReferences.push(run.referenceString)
  }

  // Everything is :+1:
  const installationSettings = {
    iID: settings.installationID,
    settings: settings.installationSettings,
  }

  let dangerDSL: DangerDSLType
  try {
    dangerDSL = await createPRDSL(githubAPI)
  } catch (error) {
    const message = `Not running Danger because we could not access the GitHub PR ${error}.`
    return {
      fails: [],
      markdowns: [],
      warnings: [],
      messages: [{ message }],
    }
  }

  const results = await runDangerForInstallation(
    eventName,
    contents,
    dangerfileReferences,
    githubAPI,
    RunType.pr,
    installationSettings,
    {
      dsl: dangerDSL,
      webhook: null,
    }
  )

  return results ? results : null
}

const validateRuns = async (
  runs: DangerRun[],
  settings: GitHubRunSettings,
  token: string,
  pr: Pull_request
): Promise<DangerResults | null> => {
  // In theory only a PR requires a custom branch, so we can check directly for that
  // in the event JSON and if it's not there then use master
  // prioritise the run metadata

  const dangerfileRepoForPR = pr.head.repo.full_name
  const dangerfileBranchForPR = pr.head.ref

  for (const run of runs) {
    const neededDangerfileIsLocalRepo = !run.repoSlug
    const branch = neededDangerfileIsLocalRepo ? dangerfileBranchForPR : null

    // Either it's dictated in the run as an external repo, or we use the most natural repo
    const repoForDangerfileRun = run.repoSlug || dangerfileRepoForPR

    // TODO: This can't be right, they're the same branch/repo reference?
    const baseDangerfile = await getGitHubFileContents(token, repoForDangerfileRun, run.dangerfilePath, branch)
    const headDangerfile = await getGitHubFileContents(token, repoForDangerfileRun, run.dangerfilePath, branch)
    const dangerfilesExist = headDangerfile !== "" && baseDangerfile !== ""

    // Shortcut to determine if both Dangerfile exists, and that they have different content
    if (dangerfilesExist && baseDangerfile !== headDangerfile) {
      // Check to see if they have write access, if they don't then don't run the
      // Dangerfile, but put out a message that it's not being ran on purpose
      const userCanWrite = await canUserWriteToRepo(token, settings.triggeredByUsername!, dangerfileRepoForPR)
      if (!userCanWrite) {
        const message = "Not running Danger rules due to user with no write access changing the Dangerfile."
        return {
          fails: [],
          markdowns: [],
          warnings: [],
          messages: [{ message }],
        }
      }
    }

    const reportData = (reason: string) => {
      const stateForErrorHandling = {
        branch,
        dangerfileBranchForPR,
        neededDangerfileIsLocalRepo,
        repoForDangerfileRun,
        run,
        settings,
      }

      return `${reason}
  
  ## Full state of PR run:
  
  \`\`\`json
  ${JSON.stringify(stateForErrorHandling, null, "  ")}
  \`\`\`
        `
    }

    if (headDangerfile === "") {
      const actualBranch = branch ? branch : "master"
      const message = `Could not find Dangerfile at <code>${
        run.dangerfilePath
      }</code> on <code>${repoForDangerfileRun}</code> on branch <code>${actualBranch}</code>`

      const report = reportData(message)
      return {
        fails: [{ message: report }],
        markdowns: [],
        warnings: [],
        messages: [],
      }
    }
  }
  return null
}
