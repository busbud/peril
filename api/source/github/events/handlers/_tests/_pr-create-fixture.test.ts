import { format } from "prettier"

jest.mock("../../../../db/getDB")
import { MockDB } from "../../../../db/__mocks__/getDB"
import { getDB } from "../../../../db/getDB"
const mockDB = getDB() as MockDB

jest.mock("../../../../runner/runFromSameHost")
import { runFromSameHost } from "../../../../runner/runFromSameHost"

jest.mock("../../../../api/github", () => ({
  getTemporaryAccessTokenForInstallation: () => Promise.resolve("12345"),
}))

jest.mock("../../../../github/lib/github_helpers", () => ({
  getGitHubFileContents: jest.fn(),
  canUserWriteToRepo: () => true,
  doesPRExist: () => Promise.resolve(true),
}))
import { getGitHubFileContents } from "../../../lib/github_helpers"
const mockGetGitHubFileContents: any = getGitHubFileContents

import { readFileSync, writeFileSync } from "fs"
import { resolve } from "path"
import { dangerRunForRules } from "../../../../danger/danger_run"

import { PerilRunnerBootstrapJSON } from "../../../../runner/triggerSandboxRun"
import { setupForRequest } from "../../github_runner"

import { DangerResults } from "danger/distribution/dsl/DangerResults"
import { FakePlatform } from "danger/distribution/platforms/FakePlatform"
import { runPRRun } from "../pr"

const mockPlatform = new FakePlatform()
jest.mock("../../../../danger/peril_platform", () => ({
  getPerilPlatformForDSL: () => mockPlatform,
}))

jest.mock("../../createPRDSL", () => ({ createPRDSL: jest.fn() }))
import { createPRDSL } from "../../createPRDSL"

const apiFixtures = resolve(__dirname, "../../_tests/fixtures")
const fixture = (file: string) => JSON.parse(readFileSync(resolve(apiFixtures, file), "utf8"))

it("passes the right args to the hyper functions when it's a PR", async () => {
  ;(createPRDSL as jest.Mock).mockResolvedValue({})

  mockDB.getInstallation.mockReturnValue({ iID: "123", repos: {}, envVars: { hello: "world" } })

  const body = fixture("pull_request_opened.json")
  const req = { body, headers: { "X-GitHub-Delivery": "123" } } as any
  const settings = await setupForRequest(req, {})

  const dangerfileForRun = "warn(danger.github.api)"
  mockGetGitHubFileContents.mockImplementationOnce(() => Promise.resolve(dangerfileForRun))

  const runSettings = { pull_request: "danger/peril-settings@testing/logger.ts" }
  const run = dangerRunForRules("pull_request", "opened", runSettings, body)

  await runPRRun("eventName", run, settings, "12345", body.pull_request)

  // Take the payload, remove the JWT and save a copy of the JSON into a fixture dir, then snapshot it
  const payload = (runFromSameHost as any).mock.calls[0][0] as PerilRunnerBootstrapJSON
  payload.perilSettings.perilJWT = "12345"
  payload.perilSettings.perilRunID = "[run-id]"

  const file = format(JSON.stringify(payload, null, "  "), { parser: "json" })
  writeFileSync(__dirname + "/fixtures/PerilRunnerPRBootStrapExample.json", file, "utf8")

  expect(payload).toMatchSnapshot()
})

it("catches error from missing PR", async () => {
  ;(createPRDSL as jest.Mock).mockRejectedValue({})
  mockDB.getInstallation.mockReturnValue({ iID: "123", repos: {}, envVars: { hello: "world" } })

  const body = fixture("pull_request_opened.json")
  const req = { body, headers: { "X-GitHub-Delivery": "123" } } as any
  const settings = await setupForRequest(req, {})

  const dangerfileForRun = "warn(danger.github.api)"
  mockGetGitHubFileContents.mockImplementationOnce(() => Promise.resolve(dangerfileForRun))

  const runSettings = { pull_request: "danger/peril-settings@testing/logger.ts" }
  const run = dangerRunForRules("pull_request", "opened", runSettings, body)

  expect.hasAssertions()
  const results = await runPRRun("eventName", run, settings, "12345", body.pull_request)

  expect(results).not.toBeNull()
  expect((results as DangerResults).messages.length).toBeGreaterThan(0)
})
