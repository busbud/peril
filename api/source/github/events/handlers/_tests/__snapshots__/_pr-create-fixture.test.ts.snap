// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the right args to the hyper functions when it's a PR 1`] = `
Object {
  "dslType": "pr",
  "installation": Object {
    "iID": 4766,
    "settings": Object {},
  },
  "paths": Array [
    "danger/peril-settings@testing/logger.ts",
  ],
  "payload": Object {
    "dsl": Object {
      "settings": Object {
        "cliArgs": Object {},
        "github": Object {
          "accessToken": "12345",
          "additionalHeaders": Object {
            "Accept": "application/vnd.github.machine-man-preview+json",
          },
          "baseURL": undefined,
        },
      },
    },
    "webhook": null,
  },
  "perilSettings": Object {
    "envVars": Object {
      "hello": "world",
    },
    "event": "eventName",
    "perilAPIRoot": undefined,
    "perilJWT": "12345",
    "perilRunID": "[run-id]",
  },
}
`;
